src/Extend/Model.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Extend;
+
+use Flarum\Extension\Extension;
+use Illuminate\Contracts\Container\Container;
+
+class Model implements ExtenderInterface
+{
+    private $modelClass;
+    private $dateCallbacks = [];
+    private $defaultAttributeCallbacks = [];
+    private $relationships = [];
+
+    /**
+     * @param string $modelClass The ::class attribute of the model you are modifying.
+     *                           This model should extend from \Flarum\Database\AbstractModel.
+     */
+    public function __construct(string $modelClass)
+    {
+        $this->modelClass = $modelClass;
+    }
+
+    /**
+     * Modify which attributes of this model are treated as dates.
+     *
+     * @param callable $callable
+     *
+     * The callable can be a closure or invokable class, and should accept:
+     * - \Flarum\User\User $user: the user in question.
+     * - array $dateAttributes: an array of attributes which are processed as dates.
+     *
+     * The callable should return:
+     * - array $dateAttributes: an array of attributes which are processed as dates.
+     */
+    public function configureDates(callable $callable)
+    {
+        $this->dateCallbacks[] = $callable;
+
+        return $this;
+    }
+
+    /**
+     * Modify default attribute values for this model.
+     *
+     * @param callable $callable
+     *
+     * The callable can be a closure or invokable class, and should accept:
+     * - \Flarum\User\User $user: the user in question.
+     * - array[string] $dateAttributes: an associative array of attributes => default values.
+     *
+     * The callable should return:
+     * - array[string] $dateAttributes: an associative array of attributes => default values.
+     */
+    public function configureDefaultAttributes(callable $callable)
+    {
+        $this->defaultAttributeCallbacks[] = $callable;
+
+        return $this;
+    }
+
+    /**
+     * Add a relationship from this model to another model.
+     *
+     * @param string $name: the name of the relation. This doesn't have to match anything,
+     *                      but has to be unique from other relation names for this model.
+     * @param callable $callable
+     *
+     * The callable can be a closure or invokable class, and should accept:
+     * - $instance: An instance of this model.
+     *
+     * The callable should return:
+     * - $relationship: A Laravel Relationship object. See relevant methods of models
+     *                  like \Flarum\User\User for examples of how relationships should be returned.
+     */
+    public function relationship(string $name, callable $callable)
```

Comment:
For these ones, I think that accepting the arguments here directly makes using these extenders much nicer.

We'd have two options:
~~~php
->relationship('posts', 'hasMany', ...)

->hasMany('posts', ...)
~~~

Personally, I prefer the second option, as that's even less typing and the list is not very likely to get any bigger. If you experiment a bit with [variadic arguments](https://www.php.net/manual/en/functions.arguments.php#functions.variable-arg-list), I guess you won't even have to explicitly pass on all the different arguments for each type. (Except for listing them in the docblock, I guess.)

Reply:
This would further reduce the boilerplate for certain relationship types, by the way, because we would not always to specify [lots of default values just to set the last argument](https://github.com/flarum/mentions/blob/9ac08e302d01849a6252fe4339c8eb1b415a12b3/src/Listener/AddPostMentionedByRelationship.php#L40), which the extender could derive from the first argument (the relationship name).

Reply:
I'm going to keep the main relationship one that accepts a callback for edge cases, and provide specific ones (that use the relationship one) for the main cases.

Reply:
Done.